File: sum_root_to_leaf_numbers_iterative.py
"""
Problem: 129. Sum Root to Leaf Numbers
Leetcode: https://leetcode.com/problems/sum-root-to-leaf-numbers/
Solution: Iterative DFS preorder traversal
Time Complexity: O(n), we have to traverse all numbers
Space Complexity: O(h), h is the tree height, recursive approach uses inbuilt stack to store all numbers
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumNumbers(self, root: TreeNode):
        total_sum = 0
        stack = [(root, 0)]

        while stack:
            root, curr_number = stack.pop()
            if root:
                curr_number = curr_number * 10 + root.val
                # if it's a leaf, update total_sum
                if not root.left and not root.right:
                    total_sum += curr_number
                else:
                    # push right first and then left so that left can be popped first as we're doing preorder traversal
                    stack.append((root.right, curr_number))
                    stack.append((root.left, curr_number))
        return total_sum

File: SumRootToLeafNumbersIterative.java
// Time complexity: O(N)
// Space complexity: O(H)

// Iterative approach

class Pair {
    TreeNode root;
    int currSum;
    
    public Pair(TreeNode root, int currSum) {
        this.root = root;
        this.currSum = currSum;
    }
}
class Solution {
    public int sumNumbers(TreeNode root) {
        Stack<Pair> stack = new Stack<>();
        int currSum = 0;
        int result = 0;
        while (root != null || !stack.isEmpty()) {
            while (root != null) {
                currSum = currSum*10 + root.val;
                stack.push(new Pair(root, currSum));
                root = root.left;
            }
            Pair pair = stack.pop();
            root = pair.root;
            currSum = pair.currSum;
            if (root.left == null && root.right == null) {
                result += currSum;
            }
            root = root.right;
        }
        return result;
    }
}